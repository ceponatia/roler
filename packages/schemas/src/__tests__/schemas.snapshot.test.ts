import { describe, expect, it } from 'vitest';

import {
  CharacterCanonicalSchema,
  CharacterInstanceSchema,
  SceneInstanceSchema,
  RelationshipInstanceSchema,
  MemoryInstanceSchema,
  RelationshipTypeCanonicalSchema
} from '../index.js';

import type { ZodTypeAny } from 'zod';

// Collect schemas to snapshot their .describe() output for surface stability.
const schemaEntries = [
  ['CharacterCanonical', CharacterCanonicalSchema],
  ['CharacterInstance', CharacterInstanceSchema],
  ['SceneInstance', SceneInstanceSchema],
  ['RelationshipInstance', RelationshipInstanceSchema],
  ['MemoryInstance', MemoryInstanceSchema],
  ['RelationshipTypeCanonical', RelationshipTypeCanonicalSchema]
] as const;

// Zod's .describe(description: string) mutator requires an argument; we want a read-only
// structural snapshot. This helper extracts a small, stable subset of internal metadata.
const describeSchema = (schema: ZodTypeAny) => {
  // Access _def via any to avoid depending on private typings; kept deliberately shallow
  const def = (schema as any)._def ?? {};
  const typeName = def.typeName;
  const description = def.description;
  // For object schemas capture sorted shape keys (excluding functions)
  let shapeKeys: readonly string[] | undefined;
  try {
    const shape = typeof def.shape === 'function' ? def.shape() : def.shape;
    if (shape && typeof shape === 'object' && !Array.isArray(shape)) {
      shapeKeys = Object.keys(shape).sort();
    }
  } catch {
    /* ignore */
  }
  // For unions capture option type names
  let options: readonly string[] | undefined;
  if (Array.isArray(def.options)) {
    options = def.options.map((o: any) => o?._def?.typeName).filter(Boolean).sort();
  }
  return { typeName, description, shapeKeys, options };
};

describe('schema descriptions', () => {
  for (const [name, schema] of schemaEntries) {
    it(`matches snapshot: ${name}`, () => {
      expect(describeSchema(schema)).toMatchSnapshot();
    });
  }
});
