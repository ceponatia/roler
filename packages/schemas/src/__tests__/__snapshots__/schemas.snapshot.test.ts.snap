// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`schema descriptions > matches snapshot: CharacterCanonical 1`] = `
{
  "description": undefined,
  "options": undefined,
  "shapeKeys": [
    "ageCheck",
    "attributes",
    "bio",
    "blockedTags",
    "contentRating",
    "createdAt",
    "createdBy",
    "id",
    "kind",
    "lineageId",
    "name",
    "source",
    "status",
    "updatedAt",
    "updatedBy",
    "version",
  ],
  "typeName": "ZodObject",
}
`;

exports[`schema descriptions > matches snapshot: CharacterInstance 1`] = `
{
  "description": undefined,
  "options": undefined,
  "shapeKeys": [
    "attributes",
    "canonicalId",
    "changePolicy",
    "createdAt",
    "createdBy",
    "id",
    "kind",
    "mutableAttributes",
    "name",
    "sessionId",
    "updatedAt",
    "updatedBy",
  ],
  "typeName": "ZodObject",
}
`;

exports[`schema descriptions > matches snapshot: MemoryInstance 1`] = `
{
  "description": undefined,
  "options": undefined,
  "shapeKeys": [
    "attributes",
    "canonicalId",
    "changePolicy",
    "createdAt",
    "createdBy",
    "decayPolicy",
    "id",
    "kind",
    "mutableAttributes",
    "salience",
    "scope",
    "updatedAt",
    "updatedBy",
    "vector",
  ],
  "typeName": "ZodObject",
}
`;

exports[`schema descriptions > matches snapshot: RelationshipInstance 1`] = `
{
  "description": undefined,
  "options": undefined,
  "shapeKeys": [
    "attributes",
    "boundaries",
    "canonicalId",
    "changePolicy",
    "consentFlags",
    "createdAt",
    "createdBy",
    "history",
    "id",
    "intensity",
    "kind",
    "mutableAttributes",
    "participants",
    "romance",
    "stage",
    "typeRef",
    "updatedAt",
    "updatedBy",
  ],
  "typeName": "ZodObject",
}
`;

exports[`schema descriptions > matches snapshot: RelationshipTypeCanonical 1`] = `
{
  "description": undefined,
  "options": undefined,
  "shapeKeys": [
    "ageCheck",
    "allowedParticipants",
    "attributes",
    "blockedTags",
    "contentRating",
    "createdAt",
    "createdBy",
    "defaultStages",
    "exclusivity",
    "id",
    "intimacyEscalationAllowed",
    "kind",
    "lineageId",
    "name",
    "romanceEnabled",
    "source",
    "status",
    "updatedAt",
    "updatedBy",
    "version",
  ],
  "typeName": "ZodObject",
}
`;

exports[`schema descriptions > matches snapshot: SceneInstance 1`] = `
{
  "description": undefined,
  "options": undefined,
  "shapeKeys": [
    "attributes",
    "beats",
    "canonicalId",
    "changePolicy",
    "createdAt",
    "createdBy",
    "id",
    "kind",
    "mutableAttributes",
    "participants",
    "sessionId",
    "settingRef",
    "summary",
    "tags",
    "title",
    "updatedAt",
    "updatedBy",
  ],
  "typeName": "ZodObject",
}
`;
